paxos/node.cc

```diff
@@ -77,30 +78,86 @@ void PaxosNode::TruncationLoop(std::stop_token st) {
 
   while (!st.stop_requested()) {
     // TODO(mmucklo): fill with Truncation logic.
-    size_t acceptor_stubs_size;
-    {
-      absl::ReaderMutexLock l(&lock_);
-      acceptor_stubs_size = acceptor_stubs_.size();
-    }
-    for (size_t i = 0; i < acceptor_stubs_size; i++) {
-      std::unique_ptr<paxos_rpc::Acceptor::Stub>& stub = GetAcceptorStub(i);
-      paxos_rpc::TruncateProposeRequest request;
-      paxos_rpc::TruncateProposeResponse response;
-      grpc::ClientContext context;
-      absl::ReaderMutexLock l(&lock_);
-      if (stub == nullptr) {
-        // We can't truncate without agreement. Since a node is down, we skip
-        // this round.
-        LOG(INFO) << "Skipping truncation because node: " << i
-                  << " is offline.";
-        break;
+    if (IsLeader()) {
+      // What happens if two truncation loops run on two different nodes
+      // such as during a leader flip. This is okay. We will just enqueue
+      // two truncations at worse. We can optimize this edge case later,
+      // which would only add potentially extra RPCs during leader moves.
+      size_t acceptor_stubs_size;
+      {
+        absl::ReaderMutexLock l(&lock_);
+        acceptor_stubs_size = acceptor_stubs_.size();
       }
+      uint64_t min_index = std::numeric_limits<uint64_t>::max();
+      bool skip = false;
+      for (size_t i = 0; i < acceptor_stubs_size; i++) {
+        if (!IsLeader()) {
+          break;
+        }  // small optimization, but as explained above still okay if we don't
```

Comment:
Looks like the comment is in the wrong place ?

Reply:
fixed.

Reply:
fixed.